File: nexus/ts/hello-world/src/caller/workflows.ts
import * as wf from '@temporalio/workflow';
import { NexusOperationFailure } from '@temporalio/common';
import { sayHelloService } from '../api';

const ENDPOINT = 'hello-nexus-basic-nexus-endpoint';

export const callNexusSignal = wf.defineSignal<[string]>('callNexus');
export const finishSignal = wf.defineSignal('finish');
export const getResultsQuery = wf.defineQuery<string[]>('getResults');

export async function syncCallerWorkflow(name: string): Promise<string> {
  const client = wf.createNexusServiceClient({ service: sayHelloService, endpoint: ENDPOINT });
  const result = await client.executeOperation('mySyncOperation', { name }, { scheduleToCloseTimeout: '10s' });
  return result.message;
}

// Long-running workflow that calls the async nexus operation each time it receives a callNexus signal.
// Useful for exploring what happens when the handler workflow (started by the nexus operation) is reset.
// Send 'finish' signal to complete the workflow and get results.
export async function asyncCallerWorkflow(): Promise<string[]> {
  const client = wf.createNexusServiceClient({ service: sayHelloService, endpoint: ENDPOINT });
  const pendingNames: string[] = [];
  const results: string[] = [];
  let finished = false;

  wf.setHandler(callNexusSignal, (name) => {
    pendingNames.push(name);
  });

  wf.setHandler(finishSignal, () => {
    finished = true;
  });

  wf.setHandler(getResultsQuery, () => results);

  while (!finished) {
    await wf.condition(() => pendingNames.length > 0 || finished);
    if (pendingNames.length > 0) {
      const name = pendingNames.shift()!;
      try {
        const result = await client.executeOperation('sayHello', { name }, { scheduleToCloseTimeout: '30s' });
        results.push(result.message);
        wf.log.info('Nexus call completed', { name, message: result.message });
      } catch (err: unknown) {
        if (err instanceof NexusOperationFailure) {
          // Non-retryable: arrives immediately after the handler workflow fails.
          // Retryable: arrives after scheduleToCloseTimeout is exceeded.
          const cause = err.cause instanceof Error ? err.cause.message : String(err.cause);
          const msg = `[FAILED] ${err.operation}: ${cause}`;
          wf.log.error('Nexus async operation failed', { name, operation: err.operation });
          results.push(msg);
        } else {
          throw err;
        }
      }
    }
  }

  return results;
}

File: nexus/python/hello-world/app/caller/workflows.py
from datetime import timedelta

from temporalio import workflow
from temporalio.exceptions import NexusOperationError

with workflow.unsafe.imports_passed_through():
    from app.service import MyInput, SayHelloService, MyOutput

NEXUS_ENDPOINT = "hello-nexus-basic-nexus-endpoint"


# This is a workflow that calls two nexus operations.
@workflow.defn
class CallerWorkflow:
    # An __init__ method is always optional on a workflow class. Here we use it to set the
    # nexus client, but that could alternatively be done in the run method.
    def __init__(self):
        self.nexus_client = workflow.create_nexus_client(
            service=SayHelloService,
            endpoint=NEXUS_ENDPOINT,
        )

    # The workflow run method invokes two nexus operations.
    @workflow.run
    async def run(self, name: str) -> tuple[MyOutput, MyOutput]:
        # Start the nexus operation and wait for the result in one go, using execute_operation.
        op_1_result = await self.nexus_client.execute_operation(
            SayHelloService.my_sync_operation,
            MyInput(name=name),
        )
        # Alternatively, you can use start_operation to obtain the operation handle and
        # then `await` the handle to obtain the result.
        op_2_handle = await self.nexus_client.start_operation(
            SayHelloService.say_hello,
            MyInput(name=name),
            schedule_to_close_timeout=timedelta(seconds=30),
        )
        try:
            op_2_result = await op_2_handle
        except NexusOperationError as e:
            # Non-retryable: arrives immediately after the handler workflow fails.
            # Retryable: arrives after schedule_to_close_timeout is exceeded.
            workflow.logger.error(
                "Nexus async operation failed",
                extra={"operation": e.operation, "service": e.service, "cause": str(e.cause)},
            )
            return op_1_result, MyOutput(message=f"[FAILED] {e.operation}: {e.cause}")
        return op_1_result, op_2_result

        # return op_2_result
